src/light_data.c

```diff
@@ -1495,27 +1493,89 @@ void light_set_lights_on(char state)
+      shadow_end = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x, subtile_coord(stl_y, 0) - pos_y) & LbFPMath_AngleMask;
+    }
+    else
+    {
+      shadow_start = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x - 1, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+      shadow_end = LbArcTanAngle(subtile_coord(stl_x + 1, 0) - pos_x, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+    }
+  }
+  else if ( y == stl_y )
+  {
+    if ( x <= stl_x )
+    {
+      shadow_start = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x, subtile_coord(stl_y, 0) - pos_y) & LbFPMath_AngleMask;
+      shadow_end = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+    }
+    else
+    {
+      shadow_start = LbArcTanAngle(subtile_coord(stl_x + 1, 0) - pos_x, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+      shadow_end = LbArcTanAngle(subtile_coord(stl_x + 1, 0) - pos_x, subtile_coord(stl_y, 0) - pos_y) & LbFPMath_AngleMask;
+    }
+  }
+  else
+  {
+    switch ( a3 )
+    {
+      case 1:
+        shadow_start = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x, subtile_coord(stl_y, 0) - pos_y) & LbFPMath_AngleMask;
+        shadow_end = LbArcTanAngle(subtile_coord(stl_x + 1, 0) - pos_x, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+        break;
+      case 2:
+        shadow_start = LbArcTanAngle(subtile_coord(stl_x + 1, 0) - pos_x, subtile_coord(stl_y, 0) - pos_y) & LbFPMath_AngleMask;
+        shadow_end = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+        break;
+      case 3:
+        shadow_start = LbArcTanAngle(subtile_coord(stl_x + 1, 0) - pos_x, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+        shadow_end = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x, subtile_coord(stl_y, 0) - pos_y) & LbFPMath_AngleMask;
+        break;
+      case 4:
+        shadow_start = LbArcTanAngle(subtile_coord(stl_x, 0) - pos_x, subtile_coord(stl_y + 1, 0) - pos_y) & LbFPMath_AngleMask;
+        shadow_end = LbArcTanAngle(subtile_coord(stl_x + 1, 0) - pos_x, subtile_coord(stl_y, 0) - pos_y) & LbFPMath_AngleMask;
+        break;
+      default:
+        shadow_end = shadow_start;
+        break;
+    }
+  }
+  if ( (shadow_start / 512) << 9 != shadow_start )
+    shadow_start = (shadow_start + 1) & 0x7FF;
+  if ( (shadow_end / 512) << 9 != shadow_end )
+    shadow_end = (shadow_end - 1) & 0x7FF;
+  result = shadow_start;
+  *shadow_limit_idx_start = shadow_start;
+  *shadow_limit_idx_end = shadow_end;
+  return result;
 }
 
 //sub_408530
 static TbBool light_render_light_sub1_sub2(MapSubtlCoord stl_x, MapSubtlCoord stl_y, MapSubtlCoord stl_z)
 {
-  return _DK_light_render_light_sub1_sub2(stl_x, stl_y, stl_z);
-
+    struct Column *col = get_column_at(stl_x, stl_y);
+    return ((col->bitfields >> 4) > stl_z);
```

Comment:
col->bitfields >> 4 could be replaced with get_column_floor_filled_subtiles

Reply:
What about the 0xF0?

Reply:
since the bits that are hidden by the mask would be dropped by the bit shift anyways the mask is technically redundant, but still a bit more readable with the mask still there